packages/react/popper/src/Popper.tsx

```diff
@@ -18,17 +18,78 @@ import type { Measurable } from '@radix-ui/rect';
 const POPPER_NAME = 'Popper';
 
 type PopperContextValue = {
+  anchorRef: React.MutableRefObject<Measurable | null>;
+};
+const [PopperProvider, usePopperContext] = createContext<PopperContextValue>(POPPER_NAME);
+
+const Popper: React.FC = ({ children }) => {
+  const anchorRef = React.useRef<Measurable | null>(null);
+  return <PopperProvider anchorRef={anchorRef}>{children}</PopperProvider>;
+};
+
+Popper.displayName = POPPER_NAME;
+
+/* -------------------------------------------------------------------------------------------------
+ * PopperAnchor
+ * -----------------------------------------------------------------------------------------------*/
+
+const ANCHOR_NAME = 'PopperAnchor';
+
+type PopperAnchorOwnProps = Polymorphic.Merge<
+  Polymorphic.OwnProps<typeof Primitive>,
+  { virtualRef?: React.RefObject<Measurable> }
+>;
+type PopperAnchorPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof Primitive>,
+  PopperAnchorOwnProps
+>;
+
+const PopperAnchor = React.forwardRef((props, forwardedRef) => {
+  const { virtualRef, children, ...anchorProps } = props;
+  const context = usePopperContext(ANCHOR_NAME);
+  const ref = React.useRef<React.ElementRef<typeof Primitive>>(null);
+  const composedRefs = useComposedRefs(
+    forwardedRef,
+    ref,
+    context.anchorRef as React.MutableRefObject<React.ElementRef<typeof Primitive>>
```

Comment:
Why do we need to cast here?

Reply:
Because it is moaning that `Measurable` is not a `HTMLDivElement`:

> Types of parameters 'node' and 'instance' are incompatible.
>          Type 'HTMLDivElement | null' is not assignable to type 'Measurable'.

This was also why I had tried doing the `T extends Measurable` thing but it didn't work haha. I'll take another look at this but it's changed anyway in the most recent PR.

Reply:
Oh wait, that's because `null` isn't assignable to `Measurable` but my `anchorRef` is `Measurable | null` 🤔 

I'll take a look, i'm confused now.

Reply:
I haven't been able to figure this out quickly and tbh it's all [replaced anyway in the next PR](https://github.com/radix-ui/primitives/pull/592/files#diff-0b55611a47427be3297d5c6da97844bb21baf24057718c14a15d277630450ac2R56). 

Do you mind if we leave this one?

Reply:
Yep that's fine, I saw the next PR.